zerver/tests/test_unread.py

```diff
@@ -137,3 +137,109 @@ def test_update_all_flags(self):
+                                                    user_profile=user_profile
+        ).values_list('message__id',flat=True))
+
+        um = list(UserMessage.objects.filter(message__recipient=recipient, user_profile=user_profile))
+        for msg in um:
+            if msg.user_profile.email == "hamlet@zulip.com":
+                self.assertFalse(msg.flags.read)
+
+        events = []
+        with tornado_redirected_to_list(events):
+            result = self.client.post("/json/messages/flags", {"messages": ujson.dumps([]),
+                                                               "op": "add",
+                                                               "flag": "read",
+                                                               "stream_name": "Scotland"})
+        self.assert_json_success(result)
+
+        event = events[0]['event']
+        expected = dict(operation='add',
+                        messages=message_ids,
+                        flag='read',
+                        type='update_message_flags',
+                        all=False)
+        self.assertTrue(len([key for key in expected if expected[key] != event[key]]) == 0)
+
+        um = list(UserMessage.objects.filter(message__subject__iexact="Denmark3"))
+        for msg in um:
+            if msg.user_profile.email == "hamlet@zulip.com":
+                self.assertTrue(msg.flags.read)
+
+    def test_mark_all_in_invalid_stream_read(self):
+        self.login("hamlet@zulip.com")
+        invalid_stream_name = ""
+        result = self.client.post("/json/messages/flags", {"messages": ujson.dumps([]),
+                                                           "op": "add",
+                                                           "flag": "read",
+                                                           "stream_name": invalid_stream_name})
+        self.assert_json_error(result, 'No such stream \'\'')
+
+    def test_mark_all_in_stream_topic_read(self):
+        self.login("hamlet@zulip.com")
+
+        result = self.client.post("/json/messages/flags", {"messages": ujson.dumps([]),
+                                                           "op": "remove",
+                                                           "flag": "read",
+                                                           "topic_name": "Denmark3",
+                                                           "stream_name": "Denmark"})
+
+        self.assert_json_success(result)
+        user_profile = get_user_profile_by_email("hamlet@zulip.com")
+        stream = get_stream("Denmark", user_profile.realm)
+        recipient = get_recipient(Recipient.STREAM, stream.id)
+        message_ids = list(UserMessage.objects.filter(message__recipient=recipient,
+                                                    user_profile=user_profile,
+                                                    message__subject__iexact="Denmark3"
+        ).values_list('message__id',flat=True))
```

Comment:
Yeah, I think I just misread your original comment.

If I `send_message` to a new stream then I only need the id that
`send_message` returns so I can accomplish the same thing as before with
more readable code, correct?

On Sat, May 7, 2016, 5:33 PM Tim Abbott <notifications@github.com> wrote:

> In zerver/tests/test_unread.py
> <https://github.com/zulip/zulip/pull/745#discussion_r62423612>:
>
> > +        self.login("hamlet@zulip.com")
> > +
> > +        result = self.client.post("/json/messages/flags", {"messages": ujson.dumps([]),
> > +                                                           "op": "remove",
> > +                                                           "flag": "read",
> > +                                                           "topic_name": "Denmark3",
> > +                                                           "stream_name": "Denmark"})
> > +
> > +        self.assert_json_success(result)
> > +        user_profile = get_user_profile_by_email("hamlet@zulip.com")
> > +        stream = get_stream("Denmark", user_profile.realm)
> > +        recipient = get_recipient(Recipient.STREAM, stream.id)
> > +        message_ids = list(UserMessage.objects.filter(message__recipient=recipient,
> > +                                                    user_profile=user_profile,
> > +                                                    message__subject__iexact="Denmark3"
> > +        ).values_list('message__id',flat=True))
>
> yeah, which is all that you're using from your message_ids queries,
> right? Maybe I don't understand what you're asking...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/zulip/zulip/pull/745/files/540791bacba89cdff11d1863603c3c9e7fb7e740#r62423612>
>